browser/autocomplete/search_suggestions_query_check_utils.cc

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/autocomplete/search_suggestions_query_check_utils.h"
+
+#include <cmath>
+#include <string>
+#include <vector>
+
+#include "base/containers/contains.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_split.h"
+#include "brave/browser/autocomplete/search_suggestions_query_check_constants.h"
+#include "content/public/common/url_constants.h"
+#include "third_party/re2/src/re2/re2.h"
+#include "url/gurl.h"
+
+namespace {
+
+double GetHashProb(const std::string& query) {
+  double log_prob = 0.0;
+  int trans_c = 0;
+  std::string updated_query = query;
+  RE2::GlobalReplace(&updated_query, "[^A-Za-z0-9]", "");
+
+  for (size_t i = 0; i < updated_query.length() - 1; i++) {
+    const auto a = updated_query[i];
+    const auto b = updated_query[i + 1];
+    if (base::Contains(kProbHashChars, a) &&
+        base::Contains(kProbHashChars, b)) {
+      const auto pos1 = kProbHashChars.at(a);
+      const auto pos2 = kProbHashChars.at(b);
+      log_prob += kProbHashLogM[pos1][pos2];
+      trans_c += 1;
+    }
+  }
+
+  if (trans_c > 0) {
+    return exp(log_prob / trans_c);
+  }
+
+  return exp(log_prob);
+}
+
+bool LooksLikeHttpPass(const std::string& query) {
+  constexpr char pattern[] = "[^:]+:[^@]+@";
+  if (RE2::PartialMatch(query, pattern)) {
+    return true;
+  }
+
+  const std::vector<std::string> tokens =
+      base::SplitString(query, " ", base::WhitespaceHandling::TRIM_WHITESPACE,
+                        base::SplitResult::SPLIT_WANT_NONEMPTY);
+  for (const auto& token : tokens) {
+    if (RE2::PartialMatch(token, pattern)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool HasEmailInQuery(const std::string& query) {
+  return RE2::PartialMatch(
+      query, "[a-z0-9\\-_@]+(@|%40|%(25)+40)[a-z0-9\\-_]+\\.[a-z0-9\\-_]");
+}
```

Comment:
Should such heavy regexps be compiled once and cached (base::NoDestructor<re2::RE2> )?